software/apps/dotp/main.c

```diff
@@ -0,0 +1,187 @@
+// Copyright 2021 ETH Zurich and University of Bologna.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// Author: Marco Bertuletti, ETH Zurich
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "encoding.h"
+#include "printf.h"
+#include "runtime.h"
+#include "synchronization.h"
+
+#include "define.h"
+
+#if defined(SINGLE) || defined(SINGLE_UNROLLED)
+#include "dotp_single.h"
+#endif
+
+#if defined(PARALLEL) || defined(PARALLEL_UNROLLED)
+#include "dotp_parallel.h"
+#endif
+
+#if defined(PARALLEL_LOCAL) || defined(LOCAL_UNROLLED)
+#include "dotp_parallel_local.h"
+#endif
+
+#if defined(PARALLEL_RED0) || defined(PARALLEL_UNROLLED_RED0)
+#include "dotp_parallel_red0.h"
+#endif
+
+#if defined(PARALLEL_REDTREE) || defined(PARALLEL_UNROLLED_REDTREE)
+#include "dotp_parallel_redtree.h"
+#endif
```

Comment:
If I am not mistaken, all the header files define different functions, so you could technically merge them into a single header or always include all of them.

Or you could have all of them have the same function names and signatures and then simplify the `main.c` to include the one you want, and the remainder of the main would be the same for all of them. At the moment, you include different header files, and then you call different functions depending on the includes, so you have the duplication twice.

For me, it's fine as is, but I just wanted to mention it in case it is possible to do that.